### ISSUE TITLE
Chart Tracker ORM Implementation

### ISSUE THREAD

**Issue by MatthewVita** [openemr/openemr#509]

**Comment by kofiav** [comment]
Will start looking at this on the weekend.

**Comment by MatthewVita** [comment]
Hey @kofiav,

How's it going? Let me know what you need to get started on this!

-Matthew

**Comment by kofiav** [comment]
Hey @MatthewVita, 
Going good so far ... here's my first commit 768b9b8. Only committed the chart tracker entity object so far. I've created a skeleton version of the chart tracker service. Not sure where it has to be instantiated or how it will be used ... but I'm still reading the codebase.

**Comment by MatthewVita** [comment]
Hi Victor,

Great start!

- Here's the main file: https://github.com/openemr/openemr/blob/master/custom/chart_tracker.php
- So the `pid` is the patient id... you can `JOIN` so that you get a full blown Doctrine entity. @robertdown has begun work on this entity here: https://github.com/openemr/openemr/issues/624 (this won't hold up your work, but please note that we need to bring these together eventually)

-Matthew

**Comment by kofiav** [comment]
Hi @MatthewVita, 

Two quick questions: 

1. Am I supposed to manually edit the autoload_static.php and autoload_classmap.php files in /vendor/compose?

2. Also I am trying to use a class I created in the services namespace but I can't seem to access a method I declared there. It keeps saying "Uncaught error: call to undefined function `function-name` in /var/www/openemr/custom/chart_tracker.php

- Victor

**Comment by MatthewVita** [comment]
Hi Victor,

Great questions. http://www.open-emr.org/wiki/index.php/Composer#Migrating should be the answer to your issue.

Basically, autoload*.php files are maintained by Composer. If you wish to add a class in a namespace, you 'll want to run composer in this way to "register" it.

There are probably other ways to do it, but this is how I do it.

**Comment by MatthewVita** [comment]
BTW, @robertdown and I are currently talking about the `Patient` entity. It may not be a good option at this time (OpenEMR allows for users to add patient data columns dynamically).

Don't worry about `JOIN`ing the entities at this time (just return the id).

**Comment by kofiav** [comment]
Hi Matthew, 

Beautiful, thanks that did the trick. About the `Patient` entity, that does sound like a thorny problem, I'd be interested in seeing the solution you come up with. Once the new `Patient` entity is ready I'll definitely use it instead of the `pid`.

**Comment by MatthewVita** [comment]
@kofiav, we're a long way off, but I listed a potential solution here: https://github.com/openemr/openemr/issues/624#issuecomment-293114845.

**Comment by kofiav** [comment]
@MatthewVita, ok gotcha, I've subscribed to the discussion and will offer any suggestions if I have any.

**Comment by kofiav** [comment]
@MatthewVita I created a new branch (f40966ab3d56cf222d6694d3c30b288966c00bfd) and committed everything except for my updates to custom/chart_tracker.php. Is there any value in creating a PR just for the chart tracker entity, repository and service (so that they can live on the master branch and be accessible to others) or should the PR ideally include the whole chart tracker Doctrine implementation including the new `Patient` contributions (??)

**Comment by MatthewVita** [comment]
@kofiav sorry for my delayed response.

Just looked at your code and it's great! I would put a PR on OpenEMR with the label "needs work" and work  on the `chart_tracker.php`. When it's done, you can assign me to review it. I feel that most of the review is already done by just looking at what you've already put together, however, it needs to be tested.

As far as the `Patient` entity, I believe that is not going to be done for a while. Would suggest moving on without it for now.

**Comment by kofiav** [comment]
@MatthewVita no worries and cool, I'm gonna put up a PR right now.

**Comment by MatthewVita** [comment]
@kofiav how's this going?

**Comment by kofiav** [comment]
@MatthewVita currently working on it. I'll update you in the next couple of days (was tough carving out time past week).

**Comment by MatthewVita** [comment]
Cool. Let me know if you need a hand with anything
